Openstack Host Aggregates support

rudexi opened the pull request:
Support for Host Aggregates in Terraform.

This is useful for putting metadata on host aggregates.
For people using host aggregates for project separation (via a scheduler hint in the metadata), the hosts list information can also be used to auto-compute quotas.

```
2020/12/24 10:55:16 [DEBUG] OpenStack Response Body: { "forbidden": { "code": 403, "message": "Policy doesn't allow os_compute_api:os-aggregates:create to be performed." } }
2020/12/24 10:55:16 [DEBUG] openstack_compute_aggregate_v2.test2: apply errored, but we're indicating that via the Error pointer rather than returning it: Error creating OpenStack aggregate: Request forbidden: [POST http://192.168.0.19/compute/v2.1/os-aggregates], error message: {"forbidden": {"code": 403, "message": "Policy doesn't allow os_compute_api:os-aggregates:create to be performed."}}
2020/12/24 10:55:16 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: Error creating OpenStack aggregate: Request forbidden: [POST http://192.168.0.19/compute/v2.1/os-aggregates], error message: {"forbidden": {"code": 403, "message": "Policy doesn't allow os_compute_api:os-aggregates:create to be performed."}}
2020/12/24 10:55:16 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error creating OpenStack aggregate: Request forbidden: [POST http://192.168.0.19/compute/v2.1/os-aggregates], error message: {"forbidden": {"code": 403, "message": "Policy doesn't allow os_compute_api:os-aggregates:create to be performed."}}
2020/12/24 10:55:16 [DEBUG] New state was assigned lineage "e721581d-401c-0cfe-13c4-7c590b9119ff" TestAccAggregateDataSource: testing.go:705: Step 0 error: errors during apply: Error: Error creating OpenStack aggregate: Request forbidden: [POST http://192.168.0.19/compute/v2.1/os-aggregates], error message: {"forbidden": {"code": 403, "message": "Policy doesn't allow os_compute_api:os-aggregates:create to be performed."}} on /tmp/tf-test296941099/main.tf line 2: (source code not available)
```

It looks like Openlab doesn't allow changes in Host Aggregates.
I am not familiar with Openlab, is it something we can configure? If not, we need a way to exclude the test and only make it available from an environment variable.

rudexi commented:
I added the correct precheck to the test, so it now works properly with Openlab.

The integration testing runs as expected in my environment (Openstack Ussuri), when used with an admin account.
```
$env | grep OS_USERNAME
OS_USERNAME=admin

$make testacc TEST=./openstack TESTARGS="-run=TestAccComputeV2Aggregate -count=1"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./openstack -v -run=TestAccComputeV2Aggregate -count=1 -timeout 120m
=== RUN   TestAccComputeV2Aggregate
--- PASS: TestAccComputeV2Aggregate (3.97s)
PASS
ok      github.com/terraform-provider-openstack/terraform-provider-openstack/openstack  4.000s

$make testacc TEST=./openstack TESTARGS="-run=TestAccAggregateDataSource -count=1"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./openstack -v -run=TestAccAggregateDataSource -count=1 -timeout 120m
=== RUN   TestAccAggregateDataSource
--- PASS: TestAccAggregateDataSource (4.70s)
PASS
ok      github.com/terraform-provider-openstack/terraform-provider-openstack/openstack  4.739s
```

ozerovandrei commented on openstack/resource_openstack_compute_aggregate_v2_test.go:
```diff
@@ -0,0 +1,75 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+
+	"strconv"
```
Could you please group stdlib packages on top here also?

rudexi replied:
Done

ozerovandrei commented on openstack/resource_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,205 @@
+				Optional: true,
+			},
+			"hosts": {
+				Type:     schema.TypeList,
```
Are you sure about using `TypeList` here? Terraform will create plan difference in case of changing hosts ordering on API side. Also hosts should be unique as well as I know. What about using `TypeSet` for this argument?

rudexi replied:
Done. I didn't know `TypeSet` could be used like this, since the official documentation doesn't give many examples, so I opted for `TypeList` thinking it was the only way to provide a simple array of string as input (instead of blocks).

rudexi commented on openstack/data_source_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,91 @@
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/aggregates"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+
+	"strconv"
```
Done

rudexi commented on openstack/resource_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,205 @@
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Actually added it after the `zone` field. If I understand correctly you separate the `Computed`/`Optional` from the `Required` in resources?

ozerovandrei replied:
For the most resources and datasources we use a blank line separator between every argument/attribute block in this provider. I propose to keep it that way.

rudexi replied:
Got it, added a newline between all arguments/attributes

rudexi commented on openstack/resource_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,205 @@
+			"metadata": {
+				Type:     schema.TypeMap,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
```
Done

rudexi commented on openstack/data_source_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,91 @@
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Done

ozerovandrei replied:
I meant adding a blank separator line between all arguments: "zone", "metadata", "hosts".

ozerovandrei commented on openstack/resource_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,208 @@
+
+	if d.HasChange("hosts") {
+		oldHosts, newHosts := d.GetChange("hosts")
+		hostsToDelete := arrayDifference(oldHosts, newHosts)
```
You don't need a new function after changing type to `TypeSet` since there is a built-in function.

You can check the example of usage: https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/master/openstack/resource_openstack_keymanager_container_v1.go#L253

rudexi replied:
Thank you for the suggestion. Changed and tested.

rudexi commented:
I added some test to cover hypervisor when given one through environment variables.
Here is the result in my local environment:
```
$make fmt && make testacc TEST=./openstack TESTARGS="-run=TestAccAggregateDataSource -count=1"
gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./openstack -v -run=TestAccAggregateDataSource -count=1 -timeout 120m
=== RUN   TestAccAggregateDataSource
--- PASS: TestAccAggregateDataSource (4.68s)
=== RUN   TestAccAggregateDataSourceWithHypervisor
--- PASS: TestAccAggregateDataSourceWithHypervisor (7.37s)
PASS
ok      github.com/terraform-provider-openstack/terraform-provider-openstack/openstack  12.087s

$make fmt && make testacc TEST=./openstack TESTARGS="-run=TestAccComputeV2Aggregate -count=1"
gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./openstack -v -run=TestAccComputeV2Aggregate -count=1 -timeout 120m
=== RUN   TestAccComputeV2Aggregate
--- PASS: TestAccComputeV2Aggregate (3.56s)
=== RUN   TestAccComputeV2AggregateWithHypervisor
--- PASS: TestAccComputeV2AggregateWithHypervisor (10.02s)
PASS
ok      github.com/terraform-provider-openstack/terraform-provider-openstack/openstack  13.620s
```

Since I caught several errors on the `hosts` and `metadata` fields updates, I corrected it and made sure to catch it with tests.

ozerovandrei commented on openstack/resource_openstack_compute_aggregate_v2.go:
```diff
@@ -0,0 +1,235 @@
+// Metadata in openstack are not fully replaced with a "set"
+// operation, instead, it's only additive, and the existing
+// metadata are only removed when set to `null` value in json.
+func mapNullFix(oldMap, newMap map[string]interface{}) (output map[string]interface{}) {
```
Could you please move this function in https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/master/openstack/util.go file since we have some common utils there.

I also think that we can use another name instead of `mapNullFix` since it can be used by other resources in the future if needed. What about `mapDiffWithNilKeys` or something like that?

Apart from it this PR looks good 👍

rudexi replied:
Done. Also added a unit test for the function

ozerovandrei reviewed (approved):
LGTM

Thank you @rudexi
